### PR TITLE
Add docstrings for new stream class, update sphinx docs

### PR DESCRIPTION
**Pull request by amitschang**
Adding some docstrings for new classes and methods. Also subbing out some pages for the manual.

### CODE REVIEW

File: dplutils/pipeline/graph.py [hunk 1]
```diff
@@ -9,6 +9,15 @@ class TRM(Enum):
 
 
 class PipelineGraph(DiGraph):
+    """Graph of pipeline tasks.
+
+    This class adds convenience functionality for task pipeline handing on top
```
Comment (jamienoss) [inline dplutils/pipeline/graph.py:14]:
```suggestion
    This class adds convenience functionality for task pipeline handling on top
```

File: dplutils/pipeline/graph.py [hunk 2]
```diff
@@ -29,6 +38,8 @@ def sink_tasks(self):
         return [n for n,d in self.out_degree() if d == 0]
 
     def to_list(self):
+        """Return list representation of task iff it is a simple-path graph
```
Comment (jamienoss) [inline dplutils/pipeline/graph.py:41]:
```suggestion
        """Return list representation of task if it is a simple-path graph
```

File: dplutils/pipeline/graph.py [hunk 3]
```diff
@@ -59,7 +70,31 @@ def _sort_key(x):
...
     def walk_fwd(self, source=None, sort_key=None):
+        """Walk graph forward in breadth-first order from ``source``
+
+        This is a generator that yeilds tasks encountered as it walks along
+        edges in the forward direction, starting at ``source``, or at the set of
+        :attr:`source_tasks` if not supplied.
+
+        Args:
+          source: starting task of walk, defaults to :attr:`source_tasks`
+          sort_key: when multiple out-egdes are encountered, sort the yeilded
```
Comment (jamienoss) [inline dplutils/pipeline/graph.py:81]:
```suggestion
          sort_key: when multiple out-egdes are encountered, sort the yielded
```

File: dplutils/pipeline/graph.py [hunk 4]
```diff
@@ -59,7 +70,31 @@ def _sort_key(x):
                 yield node
 
     def walk_fwd(self, source=None, sort_key=None):
+        """Walk graph forward in breadth-first order from ``source``
+
+        This is a generator that yeilds tasks encountered as it walks along
```
Comment (jamienoss) [inline dplutils/pipeline/graph.py:75]:
```suggestion
        This is a generator that yields tasks encountered as it walks along
```

File: dplutils/pipeline/graph.py [hunk 5]
```diff
@@ -59,7 +70,31 @@ def _sort_key(x):
...
+          sort_key: when multiple out-egdes are encountered, sort the yeilded
+            tasks in order of callable `sort_key`, which should return a
+            sortable object given :class:`PipelineTask` as input.
+        """
         return self._walk(source or TRM.source, back=False, sort_key=sort_key)
 
     def walk_back(self, source=None, sort_key=None):
+        """Walk graph backward in breadth-first order from ``source``
+
+        This is a generator that yeilds tasks encountered as it walks along
```
Comment (jamienoss) [inline dplutils/pipeline/graph.py:90]:
```suggestion
        This is a generator that yields tasks encountered as it walks along
```

File: dplutils/pipeline/graph.py [hunk 6]
```diff
@@ -59,7 +70,31 @@ def _sort_key(x):
...
+        edges in the reverse direction, starting at ``source``, or at the set of
+        :attr:`sink_tasks` if not supplied.
+
+        Args:
+          source: starting task of walk, defaults to :attr:`source_tasks`
+          sort_key: when multiple in-egdes are encountered, sort the yeilded
```
Comment (jamienoss) [inline dplutils/pipeline/graph.py:96]:
```suggestion
          sort_key: when multiple in-egdes are encountered, sort the yielded
```

File: dplutils/pipeline/ray.py [hunk 7]
```diff
@@ -135,6 +135,25 @@ class RemoteTracker:
 
 
 class RayStreamGraphExecutor(StreamingGraphExecutor):
+    """Ray-based implementation of stream graph executor.
+
+    All task outputs are kept in object store and only de-serialized as needed
+    for execution, until yeilded by :meth:`run`, where they are de-serialized on
```
Comment (jamienoss) [inline dplutils/pipeline/ray.py:141]:
```suggestion
    for execution, until yielded by :meth:`run`, where they are de-serialized on
```

File: dplutils/pipeline/ray.py [hunk 8]
```diff
@@ -135,6 +135,25 @@ class RemoteTracker:
...
+    the driver.
+
+    This executor will attempt to pack the cluster, irrespective of any other
```
Comment (jamienoss) [inline dplutils/pipeline/ray.py:144]: Should somewhere mention that we not just packing but overcommitting?

File: dplutils/pipeline/ray.py [hunk 9]
```diff
@@ -135,6 +135,25 @@ class RemoteTracker:
...
+    workloads.
+
+    Note:
+      Ray cluster will be initialized with defaults upon run if it hasn't
+      already been initialized
+
+    Args:
+      ray_poll_timeout: After scheduling as many tasks as can fit, ray.wait on
```
Comment (jamienoss) [inline dplutils/pipeline/ray.py:152]: >Should somewhere mention that we not just packing but overcommitting?

Like maybe (also) here 🤷 ?

**Review by jamienoss — CHANGES_REQUESTED** [review]
👍 
Just some typos and the possibility of mentioning that resources are overcommitted.

File: dplutils/pipeline/graph.py (hunk 2 above)
Comment (amitschang) [reply dplutils/pipeline/graph.py:41]: I meant `iff`, as in "if and only if"

File: dplutils/pipeline/ray.py (hunk 8 above)
Comment (amitschang) [reply dplutils/pipeline/ray.py:144]: I don't think that is necessary, it utilizes the ray scheduler to make sure tasks are run. The mention above is just to say that it does not consider any other workloads.

**Comment by amitschang** [comment]
@jamienoss , thanks, I probably should have run a spell-checker - and I had no idea I had such a tendency to type yeilded 😆

**Review by jamienoss — APPROVED** [review]
👍 

Though, I find it important info to know that a scheduler will overcommit such that I have a better handle on resource expectations. Maybe I should just always assume that everything overcommits.

**Comment by amitschang** [comment]
Thanks. And to be clear, it doesn't overcommit in the sense that more tasks than resources run expecting that some will not use what they request. Ray will ensure that the resources requested are guaranteed for the task (e.g. if you have 8 cores and tasks using 5 and 4, both will not run at the same time)

**Comment by jamienoss** [comment]
Ok, good to know - I guess I misunderstood a comment in #43 but never mind. Maybe I'm just misremembering it lol.

**Pull request merged**